designs/2024-deprecated-rule-metadata/README.md

```diff
@@ -0,0 +1,250 @@
+    | string // Shorthand property for general deprecation message, such as why the deprecation occurred. Any truthy value implies deprecated.
+    | {
+        message?: string; // General deprecation message, such as why the deprecation occurred.
+        url?: string; // URL to more information about this deprecation in general.
+      };
+  replacedBy:
+    | readonly string[] // Existing shorthand property for rule names, backwards compatible. It's recommended to omit the plugin prefix from rule names.
+    | readonly {
+        /**
+         * Plugin containing the replacement.
+         * Use "eslint" if the replacement is an ESLint core rule.
+         * Omit if the replacement is in the same plugin.
+         */
+        plugin?:
+          | string // Shorthand property for the plugin name i.e. "eslint-plugin-example" that contains the replacement rule.
+          | {
+              name?: string; // Plugin name i.e. "eslint-plugin-example" that contains the replacement rule.
+              url?: string; // URL to plugin documentation.
+            };
+        rule:
+          | string // Shorthand property for rule name (without plugin prefix).
+          | {
+              name?: string; // Replacement rule name (without plugin prefix).
+              url?: string; // URL to rule documentation.
+            };
+        meta?: {
+          message?: string; // Message about this specific replacement, such as how to use/configure the replacement rule to achieve the same results as the rule being replaced.
+          url?: string; // URL to more information about this specific deprecation/replacement.
+        };
+      }[]
+    | undefined;
+};
+```
+
+Real-world example of how this could be used based on the situation in <https://github.com/eslint/eslint/issues/18053>:
+
+```js
+// lib/rules/semi.js
+module.exports = {
+  meta: {
+    deprecated: {
+      message: 'Stylistic rules are being moved out of ESLint core.',
+      url: 'https://eslint.org/blog/2023/10/deprecating-formatting-rules/',
+    },
+    replacedBy: [
+      {
+        plugin: {
+          name: '@stylistic/js',
+          url: 'https://eslint.style/',
+        },
+        rule: {
+          name: 'semi',
+          url: 'https://eslint.style/rules/js/semi',
+        },
+        meta: {
+          message: 'Use the `foo` option on the new rule to achieve the same behavior as before.',
+          url: 'https://example.com/how-to-migrate-to-the-new-semi-rule',
+        }
+      },
+    ],
+  },
+};
+```
+
+This data could be used by documentation websites and tooling like [eslint-doc-generator](https://github.com/bmish/eslint-doc-generator) to generate notices and links like:
+
+> semi (deprecated) \
+> Replaced by [semi](https://eslint.style/rules/js/semi) from [@stylistic/js](https://eslint.style/). \
+> Stylistic rules are being moved out of ESLint core. [Read more](https://eslint.org/blog/2023/10/deprecating-formatting-rules/).
```

Comment:
[Praise] This is a really good demonstration of the proposed change and its benefits. (and also why eslint-doc-generator is great!)

Reply:
A typical ESLint deprecation notice includes the version of ESLint that introduced the deprecation, for example "ESLint v8.53.0" for the [`semi`](https://github.com/eslint/eslint/blob/main/docs/src/rules/semi.md?plain=1#L13) rule:

> This rule was **deprecated** in ESLint v8.53.0. Please use the [corresponding rule](https://eslint.style/rules/js/semi) in [`@stylistic/eslint-plugin-js`](https://eslint.style/packages/js).

I think we want to keep that information, so we would need to store a `deprecatedSince` version key in `meta.replacedBy.info` or `meta.replacedBy.note`, depending on the name we choose, and then use it to generate the notice, if we want to automate that process.